src/preprocessing/passes/bv_abstraction.cpp

```diff
@@ -0,0 +1,67 @@
+/*********************                                                        */
+/*! \file bv_abstraction.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Mathias Preiner
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2018 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The BvAbstraction preprocessing pass
+ **
+ ** Abstract common structures over small domains to UF. This preprocessing
+ ** is particularly useful on QF_BV/mcm benchmarks and can be enabled via
+ ** option `--bv-abstraction`.
+ ** For more information see 3.4 Refactoring Isomorphic Circuits in [1].
+ **
+ ** [1] Liana Hadarean, An Efficient and Trustworthy Theory Solver for
+ **     Bit-vectors in Satisfiability Modulo Theories
+ **     https://cs.nyu.edu/media/publications/hadarean_liana.pdf
+ **/
+
+#include "preprocessing/passes/bv_abstraction.h"
+
+#include <vector>
+
+#include "options/bv_options.h"
+#include "theory/bv/abstraction.h"
+#include "theory/bv/theory_bv.h"
+#include "theory/rewriter.h"
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+using namespace CVC4::theory;
+
+BvAbstraction::BvAbstraction(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "bv-abstraction"){};
+
+PreprocessingPassResult BvAbstraction::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+  std::vector<Node> new_assertions;
+  std::vector<Node> assertions(assertionsToPreprocess->begin(),
+                               assertionsToPreprocess->end());
+  TheoryEngine* te = d_preprocContext->getTheoryEngine();
+  bv::TheoryBV* bv_theory = (bv::TheoryBV*)te->theoryOf(THEORY_BV);
+  bool changed = bv_theory->applyAbstraction(assertions, new_assertions);
+  for (unsigned i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    assertionsToPreprocess->replace(i, Rewriter::rewrite(new_assertions[i]));
+  }
+  // If we are using the lazy solver and the abstraction applies, then UF
+  // symbols were introduced.
+  if (options::bitblastMode() == bv::BITBLAST_MODE_LAZY && changed)
+  {
+    LogicRequest req(*d_preprocContext->getSmt());
+    req.widenLogic(theory::THEORY_UF);
```

Comment:
Actually, I think that it might be a good idea to add a method `PreprocessingPassContext::widenLogic()` that does this. With that, we can capture the instances where we add another theory in a preprocessing pass more explicitly and can hopefully add some feature in the future to intelligently detect issues with such a request.